python/label_pr.py

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env python3
+import argparse
+import logging
+import re
+from typing import Optional
+
+import requests
+from bs4 import BeautifulSoup
+from github import Github, Issue, Label, PullRequest
+from shared.data import get_data
+from shared.github import get_client
+from shared.log import configure_logger
+
+
+log = logging.getLogger(__name__)
+
+# region argparse
+parser = argparse.ArgumentParser(description="")
+parser.add_argument(
+    "--pr-url",
+    dest="pr_url",
+    metavar="pr-url",
+    required=True,
+    help="the URL for the PR to label and/or check for labels",
+)
+
+
+# endregion
+
+
+def get_owner_repo_num(html_url: str) -> tuple[str, str, int]:
+    """
+    Get the owner and repo name and issue/PR number from the HTML URL.
+
+    :param html_url: the URL from which to get the owner and repo name and issue/PR num
+    :return: the owner and repo name and the issue/PR number
+    :raise: ``ValueError``, if the data URL does not match the pattern
+    """
+
+    pattern = re.compile(
+        r"https://github.com/(?P<owner>.+)/(?P<repo>.+)/(issues|pull)/(?P<num>\d+)"
+    )
+    if (match := pattern.match(html_url)) is not None:
+        return match.group("owner"), match.group("repo"), int(match.group("num"))
+    raise ValueError("Could not identify owner and repo name and issue/PR number")
+
+
+def get_issue(gh: Github, html_url: str) -> Issue:
+    """
+    Get the ``Issue`` instance from a GitHub issue's HTML URL.
+
+    :param gh: the GitHub client
+    :param html_url: the HTML URL of the issue
+    :return: the ``Issue`` instance corresponding to the given URL
+    """
+
+    org, repo, issue_num = get_owner_repo_num(html_url)
+    return gh.get_organization(org).get_repo(repo).get_issue(issue_num)
+
+
+def get_pull_request(gh: Github, html_url: str) -> PullRequest:
+    """
+    Get the ``PullRequest`` instance from a GitHub PR's HTML URL.
+
+    :param gh: the GitHub client
+    :param html_url: the HTML URL of the PR
+    :return: the ``PullRequest`` instance corresponding to the given URL
+    """
+
+    org, repo, pr_num = get_owner_repo_num(html_url)
+    return gh.get_organization(org).get_repo(repo).get_pull(pr_num)
+
+
+def get_linked_issues(url: str) -> list[str]:
+    """
+    Get the list of linked issues from the GitHub UI by parsing the HTML. This is a
+    workaround because GitHub API does not provide a reliable way to find the linked
+    issues for a PR.
+
+    :param url: the URL to the GitHub UI for a PR
+    :return: the list of HTML URLs for linked issues
+    """
+
+    res = requests.get(url)
+    if res.status_code != 200:
+        return []
+
+    soup = BeautifulSoup(res.text, "html.parser")
+    divs = soup.find_all("div", **{"class": re.compile("css-truncate my-1")})
```

Comment:
Very obscure selector, this one! I'm surprised it's reliable but it does seem to be based on my own poking around the page. Hopefully they don't change the margins :stuck_out_tongue:

Reply:
It is indeed very random, I came across this hack on a forum thread and was very skeptical that a margin selector would be specific enough (but it works perfectly so 🤷 ).